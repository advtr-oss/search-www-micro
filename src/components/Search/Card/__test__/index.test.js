/**
 * @jest-environment jsdom
 */

import React from 'react'
import { Provider } from 'react-redux'
import { act, fireEvent, render, screen, waitFor, within } from '@testing-library/react'
import user from '@testing-library/user-event'
import { GlobalStyle, ThemeProvider } from '@advtr/tidy'

import configureStore from '../../../../configureStore'
import { Card } from '../index'

const store = configureStore()

// Add theme provider here too, will look at moving this out and adding a generic one
const redux = (ui, opts) => render((
  <Provider store={store}>
    <ThemeProvider>
      <React.Fragment>
        <GlobalStyle />
        <React.Fragment>{ui}</React.Fragment>
      </React.Fragment>
    </ThemeProvider>
  </Provider>
), opts)

describe('Search.Card', function () {
  it('should render a snapshot', function () {
    const { container } = redux(<Card />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render loading dropdown on focus and typing', async function () {
    user.setup()

    const { getByLabelText, container } = redux(<Card />)

    const label = getByLabelText('Search autocomplete')
    expect(label.value).toBe('')

    await act(async () => {
      label.focus()
      await user.type(label, 'H')
    })

    expect(container.firstChild).toMatchSnapshot()
    expect(label).toHaveFocus()
  })

  it('should render search dropdown on focus and typing', async function () {
    user.setup()

    // Add a fake item for now to pass the test
    const { getByLabelText, container } = redux(<Card />)

    const label = getByLabelText('Search autocomplete')
    expect(label.value).toBe('')

    await act(async () => {
      label.focus()
      await user.type(label, 'halifa')
      await delay(1000, true) // Wait for the fake API to load
    })

    const dropdown = screen.getByRole('listbox')
    const { getAllByRole } = within(dropdown)
    const items = getAllByRole('option')

    await waitFor(() => expect(items).toHaveLength(1))

    expect(container.firstChild).toMatchSnapshot()
    expect(label).toHaveFocus()
  })

  it('should update redux state on selection', async function () {
    user.setup()

    // Add a fake item for now to pass the test
    const { getByLabelText } = redux(<Card />)

    const query = 'halifa'
    const label = getByLabelText('Search autocomplete')

    await act(async () => {
      label.focus()
      await user.type(label, query)
      await delay(1000, true) // Wait for the fake API to load
    })

    const dropdown = screen.getByRole('listbox')
    const { getAllByRole } = within(dropdown)
    const items = getAllByRole('option')

    await waitFor(() => expect(items).toHaveLength(1))
    items[0].click()

    const state = store.getState()
    expect(state.search.selected.loading).toBeFalsy()
    expect(state.search.selected.$__query).toEqual(query)
  })

  it('should clear on button press', async function () {
    user.setup()

    // Add a fake item for now to pass the test
    const { getByLabelText } = redux(<Card />)

    const query = 'halifa'
    const label = getByLabelText('Search autocomplete')

    await act(async () => {
      label.focus()
      await user.type(label, query)
      await delay(1000, true) // Wait for the fake API to load
    })

    expect(label.value).toEqual(query)

    await act(() => {
      // Only works on the SVG
      const btn = getByLabelText('input-icon')
      fireEvent.click(btn)
    })

    expect(label.value).toEqual('')
  })
})

function delay (t, val) {
  return new Promise(function (resolve) {
    setTimeout(function () {
      resolve(val)
    }, t)
  })
}
