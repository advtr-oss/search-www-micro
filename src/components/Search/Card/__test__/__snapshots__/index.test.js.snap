// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search.Card should render a snapshot 1`] = `
.c3 {
  width: 100%;
  border-style: none;
  color: #24292f;
  background-color: #ffffff;
}

.c3::-webkit-input-placeholder {
  color: #6e7781;
}

.c3::-moz-placeholder {
  color: #6e7781;
}

.c3:-ms-input-placeholder {
  color: #6e7781;
}

.c3::placeholder {
  color: #6e7781;
}

.c3:disabled {
  color: #57606a;
  background-color: #f6f8fa;
}

.c3:focus {
  outline: none;
}

.c1 {
  position: relative;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 48px;
  color: #24292f;
  background-color: #ffffff;
  box-shadow: 0 3px 6px rgba(140,149,159,0.15);
  border: none;
  border-radius: 6px;
  padding: 0 8px;
}

.c1 .c2 {
  padding: 0 8px 0 8px;
  background-color: inherit;
  font-weight: 500;
  line-height: 40px;
}

.c1:disabled {
  color: #57606a;
  background-color: #f6f8fa;
}

.c4 {
  display: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  height: 40px;
  padding-left: 8px;
  color: #24292f;
  background-color: inherit;
  border-style: none;
  border-radius: 10px;
}

.c4 > svg {
  width: 16px;
  height: 16px;
}

.c4:disabled {
  cursor: default;
}

.c4:enabled > svg:hover {
  color: #0969da;
}

.c4:disabled > svg {
  cursor: default;
  color: #57606a;
}

.c0 {
  display: grid;
  gap: 16px 0;
}

<div
  class="c0"
>
  <div
    class=""
  >
    <div
      class="c1"
    >
      <input
        aria-label="Search autocomplete"
        autocomplete="off"
        class="c2 c3 input--empty"
        placeholder="Search..."
        type="text"
        value=""
      />
      <button
        class="c4"
      >
        <svg
          aria-label="input-icon"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <line
            x1="18"
            x2="6"
            y1="6"
            y2="18"
          />
          <line
            x1="6"
            x2="18"
            y1="6"
            y2="18"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`Search.Card should render loading dropdown on focus and typing 1`] = `
.c3 {
  width: 100%;
  border-style: none;
  color: #24292f;
  background-color: #ffffff;
}

.c3::-webkit-input-placeholder {
  color: #6e7781;
}

.c3::-moz-placeholder {
  color: #6e7781;
}

.c3:-ms-input-placeholder {
  color: #6e7781;
}

.c3::placeholder {
  color: #6e7781;
}

.c3:disabled {
  color: #57606a;
  background-color: #f6f8fa;
}

.c3:focus {
  outline: none;
}

.c1 {
  position: relative;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 48px;
  color: #24292f;
  background-color: #ffffff;
  box-shadow: 0 3px 6px rgba(140,149,159,0.15);
  border: none;
  border-radius: 6px;
  padding: 0 8px;
}

.c1 .c2 {
  padding: 0 8px 0 8px;
  background-color: inherit;
  font-weight: 500;
  line-height: 40px;
}

.c1:disabled {
  color: #57606a;
  background-color: #f6f8fa;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  height: 40px;
  padding-left: 8px;
  color: #24292f;
  background-color: inherit;
  border-style: none;
  border-radius: 10px;
}

.c4 > svg {
  width: 16px;
  height: 16px;
}

.c4:disabled {
  cursor: default;
}

.c4:enabled > svg:hover {
  color: #0969da;
}

.c4:disabled > svg {
  cursor: default;
  color: #57606a;
}

.c5 > ul {
  padding: 0;
  margin: 0;
  position: absolute;
  width: 100%;
  z-index: 1000;
  top: 100%;
  display: block;
  background-color: #ffffff;
  border-radius: 6px;
  overflow: hidden;
  box-shadow: 0 3px 6px rgba(140,149,159,0.15);
}

.c5 > ul * {
  cursor: default;
}

.c5 > ul:before {
  content: attr(data-title);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 40px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0 16px;
  border-bottom: 1px solid #d0d7de;
  font-size: 12px;
  color: #24292f;
  text-transform: uppercase;
  -webkit-letter-spacing: .5px;
  -moz-letter-spacing: .5px;
  -ms-letter-spacing: .5px;
  letter-spacing: .5px;
}

.c5 > ul > li:not(:last-child) {
  border-color: var(--border);
  border-bottom: 1px solid rgba(27,31,36,0.15);
}

.c6 {
  width: 100%;
}

.c7 {
  height: 48px;
  padding: 0 15px;
  display: grid;
  grid-auto-flow: column;
  grid-template-columns: max-content;
  grid-gap: 15px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  line-height: 16px;
}

.c8 {
  -webkit-animation: bQashm 1.5s ease-in-out infinite;
  animation: bQashm 1.5s ease-in-out infinite;
  background: #57606a;
  color: transparent;
  cursor: progress;
  display: block;
  width: 100%;
  height: 100%;
}

.c9 {
  width: 24px;
  height: 24px;
}

.c10 {
  height: 24px;
  width: 100%;
}

.c0 {
  display: grid;
  gap: 16px 0;
}

<div
  class="c0"
>
  <div
    class=""
  >
    <div
      class="c1"
    >
      <input
        aria-label="Search autocomplete"
        autocomplete="off"
        class="c2 c3 input--focus"
        placeholder="Search..."
        type="text"
        value="H"
      />
      <button
        class="c4"
      >
        <svg
          aria-label="input-icon"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <line
            x1="18"
            x2="6"
            y1="6"
            y2="18"
          />
          <line
            x1="6"
            x2="18"
            y1="6"
            y2="18"
          />
        </svg>
      </button>
    </div>
  </div>
  <div
    class="c5"
  >
    <ul
      class="c6"
      data-title="Search Results"
      role="listbox"
    >
      <li
        role="option"
        tabindex="-1"
      >
        <div
          class="c7 loading-item"
        >
          <span
            class="c8 c9"
          />
          <span>
            <span
              class="c8 c10"
            >
              loading...
            </span>
          </span>
        </div>
      </li>
      <li
        role="option"
        tabindex="-1"
      >
        <div
          class="c7 loading-item"
        >
          <span
            class="c8 c9"
          />
          <span>
            <span
              class="c8 c10"
            >
              loading...
            </span>
          </span>
        </div>
      </li>
    </ul>
  </div>
</div>
`;

exports[`Search.Card should render search dropdown on focus and typing 1`] = `
.c3 {
  width: 100%;
  border-style: none;
  color: #24292f;
  background-color: #ffffff;
}

.c3::-webkit-input-placeholder {
  color: #6e7781;
}

.c3::-moz-placeholder {
  color: #6e7781;
}

.c3:-ms-input-placeholder {
  color: #6e7781;
}

.c3::placeholder {
  color: #6e7781;
}

.c3:disabled {
  color: #57606a;
  background-color: #f6f8fa;
}

.c3:focus {
  outline: none;
}

.c1 {
  position: relative;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 48px;
  color: #24292f;
  background-color: #ffffff;
  box-shadow: 0 3px 6px rgba(140,149,159,0.15);
  border: none;
  border-radius: 6px;
  padding: 0 8px;
}

.c1 .c2 {
  padding: 0 8px 0 8px;
  background-color: inherit;
  font-weight: 500;
  line-height: 40px;
}

.c1:disabled {
  color: #57606a;
  background-color: #f6f8fa;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  height: 40px;
  padding-left: 8px;
  color: #24292f;
  background-color: inherit;
  border-style: none;
  border-radius: 10px;
}

.c4 > svg {
  width: 16px;
  height: 16px;
}

.c4:disabled {
  cursor: default;
}

.c4:enabled > svg:hover {
  color: #0969da;
}

.c4:disabled > svg {
  cursor: default;
  color: #57606a;
}

.c5 > ul {
  padding: 0;
  margin: 0;
  position: absolute;
  width: 100%;
  z-index: 1000;
  top: 100%;
  display: block;
  background-color: #ffffff;
  border-radius: 6px;
  overflow: hidden;
  box-shadow: 0 3px 6px rgba(140,149,159,0.15);
}

.c5 > ul * {
  cursor: pointer;
}

.c5 > ul:before {
  content: attr(data-title);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 40px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0 16px;
  border-bottom: 1px solid #d0d7de;
  font-size: 12px;
  color: #24292f;
  text-transform: uppercase;
  -webkit-letter-spacing: .5px;
  -moz-letter-spacing: .5px;
  -ms-letter-spacing: .5px;
  letter-spacing: .5px;
}

.c5 > ul > li:not(:last-child) {
  border-color: var(--border);
  border-bottom: 1px solid rgba(27,31,36,0.15);
}

.c6 {
  width: 100%;
}

.c7 {
  height: 48px;
  padding: 0 15px;
  display: grid;
  grid-auto-flow: column;
  grid-template-columns: max-content;
  grid-gap: 15px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  line-height: 16px;
}

.c7 > svg {
  font-size: 24px;
  width: 20px;
  height: 20px;
}

.c7:hover {
  background-color: rgba(175,184,193,0.2);
}

.c8 {
  font-size: 14px;
  display: inline-block;
  margin-right: 5px;
  pointer-events: none;
}

.c9 {
  font-size: 12px;
  color: #6e7781;
  display: inline;
  pointer-events: none;
}

.c0 {
  display: grid;
  gap: 16px 0;
}

<div
  class="c0"
>
  <div
    class=""
  >
    <div
      class="c1"
    >
      <input
        aria-label="Search autocomplete"
        autocomplete="off"
        class="c2 c3 input--focus"
        placeholder="Search..."
        type="text"
        value="halifa"
      />
      <button
        class="c4"
      >
        <svg
          aria-label="input-icon"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <line
            x1="18"
            x2="6"
            y1="6"
            y2="18"
          />
          <line
            x1="6"
            x2="18"
            y1="6"
            y2="18"
          />
        </svg>
      </button>
    </div>
  </div>
  <div
    class="c5"
  >
    <ul
      class="c6"
      data-title="Search Results"
      role="listbox"
    >
      <li
        role="option"
        tabindex="-1"
      >
        <div
          class="c7"
        >
          <svg
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M21 10c0 7-9 13-9 13s-9-6-9-13a9 9 0 0 1 18 0z"
            />
            <circle
              cx="12"
              cy="10"
              r="3"
            />
          </svg>
          <span>
            <span
              class="c8"
            >
              Halifax
            </span>
            <span
              class="c9"
            >
              United Kingdom
            </span>
          </span>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
